### ISSUE TITLE
Write all output files completely even if display can't be found

### ISSUE THREAD

**Issue by hoogerheide** [bumps/bumps#74]
When running fits without the --batch flag, the .err file is not populated with the statistical analysis if the display is disconnected and throws a TCLError: Cannot connect to display.

**Comment by bmaranville** [comment]
Can you provide the command string you used to start the fit?  If you can attach a zip file of the model+data that would help me recreate the error.  I ran a simple model from an ssh session with no display connected and I didn't see the error (I just ran refl1d --store=tmp model.py)

**Comment by hoogerheide** [comment]
I can do this but try the following:
1. connect over ssh with a display connected
2. start a screen and start the fit
3. disconnect the ssh session
4. See what happens to the fit output

This might allow you to reproduce the error. I think the issue is when the display is initially present but then disappears.

**Comment by bmaranville** [comment]
If you will remind me the commands required to do step 2 I will happily try it out... I haven't used "screen" in many years.

**Comment by hoogerheide** [comment]
Sure, just type "screen" at the command prompt and start the fit inside the new shell.

Then, Ctrl+a, followed by "d" to disconnect. "Exit" to close the ssh session.

When you log back in (step 4), "screen -dr" will reload your screen. Let me know if this doesn't work.

**Comment by pkienzle** [comment]
This makes no sense. The `--batch` option does two things: it redirects stdout/stderr to a .log file and it suppresses the `pylab.show()` at the end of the program [here](https://github.com/bumps/bumps/blob/7812bc040f5f5985efb3a092535be72249ea694d/bumps/cli.py#L721-L725) after all the files have been written.

The traceback should confirm that `pylab.show()` is the problem. Can you try wrapping this in a `try...except` to see if the problem goes away?

If so, then the I'm guessing some files are not being properly closed, and they don't get flushed when bumps crashes out. This is a bug that is worth fixing; all file access should be protected in a `with open` context.

**Comment by hoogerheide** [comment]
> This makes no sense. The `--batch` option does two things: it redirects stdout/stderr to a .log file and it suppresses the `pylab.show()` at the end of the program [here]

@pkienzle This happens when *not* using the --batch flag.

**Comment by pkienzle** [comment]
Exactly. If you look at the change that the --batch flag does it turns on the pylab.show() line. So the error should be happening there. And that is well after all the files have been written. So (a) confirm from the traceback that this line is causing the problem and (b) wrap it in a try...except to see if that mitigates the problem. If (b) then we need to figure out why a file is being written but not flushed.

**Comment by pkienzle** [comment]
I suppose it is possible that the TCL error is being thrown when the plot is first composed even though it does not yet need to use the display. The full traceback (with file, line no) should tell us if this is the case.
